### PR TITLE
Buffer updates

### PR DESCRIPTION
**Pull request by slide**

### CODE REVIEW

File: Src/IronPython/Runtime/ByteArray.cs [hunk 1]
```diff
@@ -904,6 +905,20 @@ public class ByteArray : IList<byte>, ICodeFormattable, IBufferProtocol
             return new ByteArray(bytes);
         }
 
+        public static ByteArray operator +(ByteArray self, string other) {
+            if(other.Any(x => x > 0xff)) {
```
Comment (slozier) [inline Src/IronPython/Runtime/ByteArray.cs]: `MakeByteArray` below does the same check as this. Maybe we could try catch the `MakeByteArray` and throw if necessary? Or if we don't want a try catch we could add a `TryMakeByteArray` which wouldn't throw.

Also, I wonder if we should add a note that this overload is specifically a workaround for the IPy string are unicode issue.

File: Src/IronPython/Runtime/Operations/StringOps.cs [hunk 2]
```diff
@@ -164,6 +165,70 @@ public class ExtensibleString : Extensible<string>, ICodeFormattable, IStructura
...
+
+        IList<BigInteger> IBufferProtocol.GetShape(int start, int? end) {
+            if (end != null) {
+                return new[] { (BigInteger)end - start };
+            }
+            return new[] { (BigInteger)_wrapped.Length - start };
+        }
+
+        void IBufferProtocol.SetItem(int index, object value) {
+            throw PythonOps.TypeError("cannot modify read - only memory");
```
Comment (slozier) [inline Src/IronPython/Runtime/Operations/StringOps.cs]: spaces: `read - only`

File: Src/IronPython/Runtime/PythonBuffer.cs [hunk 3]
```diff
@@ -238,20 +223,16 @@ public PythonBuffer(CodeContext/*!*/ context, object @object, int offset)
...
-            return Math.Max(_size, 0);
+            return Math.Max(Size, 0);
         }
 
-        internal int Size {
-            get {
-                return _size;
-            }
-        }
+        internal int Size { get; private set; }
```
Comment (slozier) [inline Src/IronPython/Runtime/PythonBuffer.cs]: Maybe not worth the effort, but it seems like `_object` and `Size` should be readonly.

**Comment by slozier** [comment]
Can we add some tests for all this stuff?

File: Src/IronPython.Modules/_socket.cs [hunk 4]
```diff
@@ -801,10 +805,34 @@ public class socket : IWeakReferenceable {
                 + "successful completion of the Send method means that the underlying system has\n"
                 + "had room to buffer your data for a network send"
                 )]
-            public void sendall(PythonBuffer data, [DefaultParameterValue(0)] int flags) {
+            public void sendall(PythonBuffer data, int flags=0) {
                 sendallWorker(data.byteCache, flags);
             }
 
+            [Documentation("sendall(string[, flags]) -> None\n\n"
```
[the remark this answers is not in the document]
Comment (slide) [reply Src/IronPython.Modules/_socket.cs]: I'm not sure how the doc stuff works in regards to something like help(_socket.sendall), we probably don't, but I kept it there like the other overloads.

**Comment by slide** [comment]
I have another PR coming that will enable tests for this, but its part of a small rework to the testing stuff.

File: Src/IronPython/Runtime/Operations/StringOps.cs (hunk 2 above)
Comment (slide) [reply Src/IronPython/Runtime/Operations/StringOps.cs]: Thanks

**Pull request merged**